### ISSUE TITLE
Merge setup.py and setup3.py

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8606]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)| @adiroiban reported|
|-|-|
|Trac ID|trac#8606|
|Type|enhancement|
|Created|2016-07-16 03:08:25Z|

there is a lot of code duplication.... and we should write a test for the setup.py itself

maybe by refactoring it into  something like https://gist.github.com/adiroiban/5b461f3ca96828a3036c99964d1d236e

<details><summary>Searchable metadata</summary>

```
trac-id__8606 8606
type__enhancement enhancement
reporter__adiroiban adiroiban
priority__normal normal
milestone__None None
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__release_management release management
keywords__None None
time__1468638505734301 1468638505734301
changetime__1474198049187875 1474198049187875
version__None None
owner__GitHub__noreply_____ GitHub <noreply@...>
cc__radix
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_2"></a>|@rodrigc commented|
|-|-|

See:
https://github.com/twisted/twisted/pull/337

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_3"></a>|@adiroiban commented|
|-|-|

Many thanks for the changes!

Please see my changes from the 8606-rodrigc-setup-2 branch https://github.com/twisted/twisted/compare/8606-rodrigc-setup-2

I tried to refactor the setup.py to use get_setup_args to just find out that this method is already defined in dist.py

I saw that the setup args are defined both in setup.py and dist.py

I think that we should move everything to dist3.py and write test for the code from there... some parts of dist.py are already covered.

I think that is bad to have setup related declarations both in setup.py and dist.py

I would also rename dist3.py to something like 

-------

Regarding the coverage for setup.py ... I can do `coverage run setup.py` and it will report as the setup.py being covered... but the point is to have something like `coverage run trial test_setup.py` to make sure that we have tests for the setup.py code.

Please commit in this branch you suggestion for the changes in tox.ini

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_4"></a>|@rodrigc commented|
|-|-|

I don't like the direction that adiroiban is taking this in.
The patch I had was simple and got the job done in terms of
merging setup3.py and setup.py.  adiroiban is making this
more complicated than it needs to be, and is wasting time.

I'm going to ask adiroiban to stop reviewing this ticket,
and I will wait to get feedback from another Twisted developer.

In terms of tox.ini, I will not modify it, but I think it is sufficient
to do ```coverage run setup.py``` to achieve coverage.  adiroiban can do that under
a separate ticket.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_5"></a>|@adiroiban commented|
|-|-|

Please don't take it personal.

My branch was just a suggestion and not a review to this patch. This is why I have not removed the review keywords.

I have only asked you to check how I would like this cleanup to be made.

-----

I think that my suggestion is a better way for fixing the setup.py coverage issue... basically once this is merged setup.py should be never touched again in other branches :) 

I think that we are not on the same page in regard to the purpose of why we want to get the coverage report.

I think that adding `coverage run setup.py` is the wrong way of fixing the setup.py coverage problem.

--------

I am also requesting the review of this branch :) https://github.com/twisted/twisted/compare/8606-rodrigc-setup-2

Here is the coverage diff https://codecov.io/gh/twisted/twisted/compare/trunk...8606-rodrigc-setup-2

My branch is merging setup.py and setup3.py but is also cleaning some code which is no longer used.

1. I have removed admin/run-pyton3 scripts as I think that we no longer need them.

2. I have renamed dist3.py to _python3_port as it really contains just the list of ported modules.

3. I have cleaned setup.py so that we define it only in dist.py ... previous some changes were in setup.py while others were in dist.py . I think that the current branch has a better design.

4. In dist.py I have removed the functions which were declarative in nature and converted them to module level variables.

5. I have simplified the way setup() is called and the way setup args are generated... they are now generated in a single place.

6. build_scripts_twisted was removed as we no longer have scripts... this should have been done in [#8491](https://github.com/twisted/twisted/issues/8491) but I don't want to revert that merge.

7. getVersion was removed as it is not used

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_6"></a>|@rodrigc commented|
|-|-|

@adi I'm going to ask you to stop here, and move on to something else, and I will get someone else to review what I've submitted.

When I have tried to submit multiple fixes for setup.py in one PR, you told me to undo things, and now you are doing the same thing.

So, please stop, and work on something else.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_7"></a>|@adiroiban commented|
|-|-|

I think that you misunderstood my role in this ticket.

I have submitted my patch as a contributor and not as a reviewer.

My change will need to be reviewed by someone else just like for any other contribution... otherwise I would have just merged my changes without a review.
I don't pretend that my changes are perfect. This is why I am asking for a review.

When I have created this ticket my intention was to have the setup.py and setup3.py merge and to also fix the coverage issue for setup.py.

I think that my cleanup branch is working on improving the coverge for setup.py and dist.py

I don't think that is wrong to have multiple people submitting solutions for the same problem.

I don't think that is wrong to disagree on a patch submitted my another contributor.

This is nothing personal. I was commenting only on this patch in particular and to your general contribution or to you as a person.

Someone reviewing this code will consider what patch to accept.

Please correct me if I am wrong.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_8"></a>|@markrwilliams commented|
|-|-|

I'd like both these changes to land.  I appreciate rodrigc's point that adioroiban's changes address a larger set of issues than just setup/setup3 redundancy.  They're still very important -- being able to test installation-related logic is critical.

First, I'd like to merge rodrigc's changes in https://github.com/twisted/twisted/pull/337.  They are low risk and provide immediate value, and should be mergeable now that we've loosened the coverage restrictions.  Once they're rebased, I will OK their being merged by removing the "review" keyword from this ticket and ensuring it gets closed.

Then I'd like to see trunk merged into adiroiban's changes in https://github.com/twisted/twisted/compare/8606-rodrigc-setup-2, and have them attached to [#8622](https://github.com/twisted/twisted/issues/8622).  I will happily review them there so they can get merged.

Does that make sense to everyone?  Let me know by dropping a comment into this ticket.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_9"></a>|@adiroiban commented|
|-|-|

Thanks for the review.

The same changes were proposed by Craig to be merged together with the console_scripts changes.

I have not accepted them in that ticket, not because I wanted to have them merge in a separate ticket exactly as they are, but because I considered that they need more work and I did not want to block the console_scripts changes.

As a common practice in Twisted, when we change some code we make sure that those changes are covered by tests, when is possible.

When I created the ticket I have not named it "merge setup.py and setup3.py and write test" since for me is obvious that all changes should come with tests... and this ticket was a good opportunity to fix the coverage. 

Also, if you merge something without tests, there is a risk that the follow up ticket for adding the test will never be followed up... I need incentives, and I fail to see the incentives in adding tests for a code which is already merged :)

Now, if you want to accept the changes without automated tests, I will not try to block you :)

Now, if you are confident that the changes are not breaking anything and that Amber will still be able to do the release by following the documented process, I will not try to block these changes.

From the wiki:

A reviewer reviews the completed work, and provides feedback: at least one good thing about the work, at least one area that needs improvement, and a judgement as to whether the good qualities ultimately outweigh the bad, i.e. whether the branch should be merged.

-------

Now is your call what code you want to be merged and how :)

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|GitHub set owner to GitHub<br>GitHub set status to `closed`|
|-|-|

In changeset 1671a0836ce81a55b8e6de42399cdecc68bb79e0
```
#!CommitTicketReference repository="" revision="1671a0836ce81a55b8e6de42399cdecc68bb79e0"
Merge pull request #337 from twisted/8606-rodrigc-setup

Author: rodrigc
Reviewer: markrwilliams
Fixes: #8606

Integrate setup3.py functionality into setup.py
```

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_11"></a>|@hawkowl commented|
|-|-|

[mass edit] Removing review from closed tickets.
